src/Azure/Orleans.Transactions.AzureStorage/TransactionalState/AzureTableTransactionalStateStorage.cs

```diff
@@ -0,0 +1,184 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.WindowsAzure.Storage.Table;
+using Newtonsoft.Json;
+using Orleans.Transactions.Abstractions;
+
+namespace Orleans.Transactions.AzureStorage
+{
+    public class AzureTableTransactionalStateStorage<TState> : ITransactionalStateStorage<TState>
+        where TState : class, new()
+    {
+        private readonly CloudTable table;
+        private readonly string partition;
+        JsonSerializerSettings JsonSettings;
+        private readonly ILogger logger;
+
+        private KeyEntity key;
+        private List<StateEntity> states;
+
+        public AzureTableTransactionalStateStorage(CloudTable table, string partition, JsonSerializerSettings JsonSettings, ILogger<AzureTableTransactionalStateStorage<TState>> logger)
+        {
+            this.table = table;
+            this.partition = partition;
+            this.JsonSettings = JsonSettings;
+            this.logger = logger;
+        }
+
+        public async Task<TransactionalStorageLoadResponse<TState>> Load()
+        {
+            try
+            {
+                Task<KeyEntity> keyTask = ReadKey();
+                Task<List<StateEntity>> statesTask = ReadStates();
+                this.key = await keyTask;
+                this.states = await statesTask;
+                if (string.IsNullOrEmpty(this.key.ETag))
+                {
+                    return new TransactionalStorageLoadResponse<TState>();
+                }
+                TState commitedState = TryFindState(this.key.CommittedTransactionId, out TState found) ? found : new TState();
+                var pendingStates = states.Select(s => new PendingTransactionState<TState>(s.TransactionId, s.SequenceId, s.GetState<TState>(this.JsonSettings))).ToList();
```

Comment:
so pending states include committed state? is this by design or?

Reply:
There is no difference between pending and committed states entities in storage.  The commited state is simply the state indicated by the key entity.